cumulusci/tasks/robotframework/robotperflistener.py

```diff
@@ -0,0 +1,142 @@
+import os
+import json
+from statistics import mean
+import time
+
+
+class TestMetrics:
+    """Manage both REST API (predefined) metrics and also custom metrics"""
+
+    def __init__(self):
+        self.api_metrics = []
+        self.custom_metrics = {}
+
+
+class RobotPerfListener:
+    """A robot listener for performance metrics from REST APIs and other code"""
+
+    ROBOT_LISTENER_API_VERSION = 2
+
+    def __init__(self, task, verbosity):
+        self._metrics = {}
+        self.task = task
+        self.verbosity = verbosity
+
+    def start_suite(self, name, attributes=None):
+        self._current_suite = {}
+        self._metrics[name] = self._current_suite
+
+    def end_suite(self, name, attributes=None):
+        self._current_suite = None
+
+    def start_test(self, name, attributes=None):
+        self._current_test = TestMetrics()
+
+    def end_test(self, name, attributes=None):
+        totals = summarize(self._current_test.api_metrics)
+        totals.update(
+            {
+                name: metric.value
+                for name, metric in self._current_test.custom_metrics.items()
+            }
+        )
+
+        test_result = {"totals": dict(sorted(totals.items()))}
+
+        if self.verbosity >= 1:
+            test_result["traces"] = self._current_test.api_metrics
+
+        self._current_suite[name] = test_result
+        self._current_test = None
+
+    def close(self):
+        filename = os.path.join(self.outputdir, "perf.json")
+        with open(filename, "w") as f:
+            json.dump(self._metrics, f, indent=2)
+
+        # cleanup so we won't be called again and also remove GC loop
+        self.task.robot_perf_listener = None
+        self.task = None
+
+    def report(self, metrics):
+        if self._current_test:  # sometimes we are called in suite teardown
+            self._current_test.api_metrics.append(metrics)
+
+    def output_file(self, path):
+        self.outputdir = os.path.dirname(path)
+
+    def create_aggregate_metric(self, name, aggregation):
+        self._current_test.custom_metrics[name] = Aggregations[aggregation]()
+
+    def store_metric_value(self, name, value):
+        self._current_test.custom_metrics[name].reduce(value)
+
+    def create_duration_metric(self, name):
+        self._current_test.custom_metrics[name] = DurationAggregator()
+
+    def end_duration_metric(self, name):
+        metric_obj = self._current_test.custom_metrics[name]
+        metric_obj.finish()
+
+
+def summarize(test_metrics):
```

Comment:
Is there a reason why this is a standalone function rather than a method on `TestMetrics`?

Reply:
I guess only historical reasons. It predates the class. But I could make it a method. I don't care much one way or the other.

Reply:
167a3215527ac69b89236be0445247b434bcfce5